sdk/core/core/internal/az_http_internal.h

```diff
@@ -0,0 +1,188 @@
+ *   - *`AZ_ERROR_ARG`*
+ *     - `p_request` is _NULL_.
+ *     - `buffer`, `method_verb`, or `initial_url` are invalid spans (see @ref az_span_is_valid).
+ *     - `max_url_size` is less than `initial_url.size`.
+ */
+AZ_NODISCARD az_result az_http_request_init(
+    _az_http_request * p_request,
+    az_http_method method,
+    az_span url,
+    az_span headers_buffer,
+    az_span body);
+
+/**
+ * @brief Adds path to url request.
+ * For instance, if url in request is `http://example.net?qp=1` and this function is called with
+ * path equals to `test`, then request url will be updated to `http://example.net/test?qp=1`.
+ *
+ *
+ * @param p_request http request builder reference
+ * @param path span to a path to be appended into url
+ * @return AZ_NODISCARD az_http_request_builder_append_path
+ */
+AZ_NODISCARD az_result az_http_request_append_path(_az_http_request * p_request, az_span path);
+
+/**
+ * @brief Set query parameter.
+ *
+ * @param p_request HTTP request builder that holds the URL to set the query parameter to.
+ * @param name URL parameter name.
+ * @param value URL parameter value.
+ *
+ * @return
+ *   - *`AZ_OK`* success.
+ *   - *`AZ_ERROR_BUFFER_OVERFLOW`* the `URL` would grow past the `max_url_size`, should the
+ * parameter gets set.
+ *   - *`AZ_ERROR_ARG`*
+ *     - `p_request` is _NULL_.
+ *     - `name` or `value` are invalid spans (see @ref az_span_is_valid).
+ *     - `name` or `value` are empty.
+ *     - `name`'s or `value`'s buffer overlap resulting `url`'s buffer.
+ */
+AZ_NODISCARD az_result
+az_http_request_set_query_parameter(_az_http_request * p_request, az_span name, az_span value);
+
+/**
+ * @brief Add a new HTTP header for the request.
+ *
+ * @param p_request HTTP request builder that holds the URL to set the query parameter to.
+ * @param key Header name (e.g. `"Content-Type"`).
+ * @param value Header value (e.g. `"application/x-www-form-urlencoded"`).
+ *
+ * @return
+ *   - *`AZ_OK`* success.
+ *   - *`AZ_ERROR_BUFFER_OVERFLOW`* there isn't enough space in the `p_request->buffer` to add a
+ * header.
+ *   - *`AZ_ERROR_ARG`*
+ *     - `p_request` is _NULL_.
+ *     - `key` or `value` are invalid spans (see @ref az_span_is_valid).
+ *     - `key` or `value` are empty.
+ *     - `name`'s or `value`'s buffer overlap resulting `url`'s buffer.
+ */
+AZ_NODISCARD az_result
+az_http_request_append_header(_az_http_request * p_request, az_span key, az_span value);
+
+/**
+ * @brief Get the HTTP header by index.
+ *
+ * @param p_request HTTP request builder.
+ * @param index Index of the HTTP header to get from the builder.
+ * @param out_result Pointer to write the result to.
+ *
+ * @return
+ *   - *`AZ_OK`* success.
+ *   - *`AZ_ERROR_ARG`*
+ *     - `p_request` or `out_result` are _NULL_.
+ *     - `index` is out of range.
+ */
+AZ_NODISCARD az_result
+az_http_request_get_header(_az_http_request * p_request, int32_t index, az_pair * out_result);
```

Comment:
This function goes in az_http_transport.h. We also need this function:
AZ_NODISCARD az_result az_http_request_get_parts(_az_http_request *p_request, az_http_method * out_method, az_span * out_url, az_span * body);